### ISSUE TITLE
Support for reading subtitles in videos, i.e. continuous and refreshable OCR

### ISSUE THREAD

**Issue by nvaccessAuto** [nvaccess/nvda#2797]
**Reported by nvdakor on 2012-11-15 06:59**
Hi,
A number of users use a screen reader which reads subtitles on videos (for example, language subtitles for foreign film videos). Currently, NVDA does not read subtitles present in some videos, so would it be possible to allow NVDA to read them? Thanks.

**Comment by OzancanKaratas** [comment]
We can NVDA controller client API on the VLC media player. It is writing with C programming language. I want to write an API, but don't know C. If you are make an API, please send it to me or VLC developer team.

**Comment by surfer0627** [comment]
A company named CatchPlay, use Brightcove's Video Cloud service. It seems that NVDA could report captions. Here is a [5 minutes demo]()https://www.youtube.com/watch?v=8dZwya7cEaQ.

notes:

* > Captioning, is more commonly used as a service to aid deaf and hearing-impaired audiences. They are more adaptable to live broadcasts, such as news broadcasts, sports events and television shows broadcast live. Usually, captions (also called closed captions) appear as white text within a black box, appearing a second or two after being spoken.
* [Brightcove](https://www.brightcove.com/en/?_ga=1.260383822.238286707.1488696146): provide Cloud Video.
* [CatchPlay](http://www.catchplay.com/tw-en#plan-intro): provide the streaming service in Taiwan, Singapore, and Indonesia.

**Comment by leonardder** [comment]
cc @JosephSL

**Comment by ehollig** [comment]
CC @josephsl for further clerification

**Comment by mirovg** [comment]
GOM Player: F5 - Others - Accessibility - Turn on "Outputing subtitles as window titles for the voice output program". Works with NVDA. Sometimes dont work at first so I try it 2 - 3 times again, but finaly it woks corretly.

**Comment by fernando-jose-silva** [comment]
I've heard voice over demos for iphone reading subtitles on netflix.
I do not know how nvda behaves on netflix, unfortunately I am not a user of this movie platform.
But if the voice over can, if netflix provides the same features for the web interface or some application for windows 10 nvda could also offer this functionality.

**Comment by OzancanKaratas** [comment]
How can we support the Netflix website for this issue?

**Comment by Adriani90** [comment]
@florianionascu7 a romanian guy developed an addon for reading subtitles. Do you know hit profile on Github? Maybe he can contribute here by raising a pull request.

**Comment by florianionascu7** [comment]
Yes, I know his Github profile. It's: @vortex1024

**Comment by vortex1024** [comment]
Hello, I am not sure this would be accepted, since it doesn't fit in basic screen reader functionality. also, the results vary greatly depending on the quality of the video and computer performance. Any NVDA developers comments about this?

**Comment by Adriani90** [comment]
In my view such a feature, if it is optional, it fits very well into screen reading basics. In fact it would be something like reading live regions. Obviously the screen reader should read the text that appears on the screen which is visible for a sighted person. So why not reading sub titles?
cc: @feerrenrut, @michaelDCurran would you accept a coresponding pull request for such a feature?

@vortex1024 could you please post a link to the github repository of the addon?

**Comment by feerrenrut** [comment]
I think it depends on the implementation, and whether it will make maintenance harder. @vortex1024 maybe you can give us an overview of the approach?

**Comment by vortex1024** [comment]
Sure. basically, what I'm doing is I run the win 10 OCR in a while loop, in a separate thread, with a configured sleeping interval. The area to scan can be   full screen, focus, foreground or navigator object. IN full screen, I added options to percentually crop zones from the screen, both to make recognition faster and easier, and to remove certain bits of the screen to be recognised, such as the TV logo or current time of the video. I've been recently asked to make the cropping options available for the foreground object too, people seem to use it for games as well.

In order not to read the same text again and again, I use the included difflib.sequence_matcher python class to determine whether the text has changed.
I haven't published the code, as I don't think it is polished enough yet, but, if there's interest for a pull request, I can try and reorganise it to the best of my abilities, then improve it based on comments.
Thanks

**Comment by OzancanKaratas** [comment]
I think we should write a handler for the subtitle files. It finds the subtitles in the videos using the video time. I think that the OCR operation will damage the content of subtitles.

It can also be considered to write a kernel mode driver such as video intercept.

**Comment by Adriani90** [comment]
There is a discussion on the lion addon also here:
https://forum.audiogames.net/topic/33489/lion-nvda-universal-subtitle-reader-and-more/

@vortex1024 gets lot of positive user feedback for this addon, and there came up several use cases for this addon:
1. In movies with subtitles on Youtube, netflix, VLC Mediaplayer etc. Very useful also for blind deaf people and people watching movies and documentaries in multiple languages
2. In mainstream games (i.e. people manage to play Fifa and lots of other games with this addon)
3. In online conferences where people share live presentations
and possibly other areas.

Current limitations:
* Creating profiles for different websites, games, apps etc is not possible. This problem would be solved if this addon was part of NVDA
* Text recognition is not 100% accurate due to OCR limitations. But this could be solved if GDI calls could be applied (i.e. also using display model). For this the help of experienced NVDA developers is needed
* Selected text cannot be recognized by the addon, so the speech is not reacting as fast as the cursor moves and stops speaking when you move between text pieces. This is because in most cases in games the text is drawn on the screen and highlighted with a certain color if it is selected, probably related to display model. I think @leonardder has proposed some improvements to display model code in NVDA. The improvements could be applied here as well I think
* When subtitles appear under eachother the addon stops working. It seems the addon waits until one piece of text disappears and another one appears. This could also be solved by applying GDI calls or by improving the OCR calls in the code.

Maybe @jcsteh and @michaelDCurran are also interested to contribute to this. I think this feature would open a large new area for blind and visually impaired users.

**Comment by Adriani90** [comment]
The first beta version of this addon is here:
http://vortex.go.ro/api/download/get/1

@vortex1024 if you could upload it to your github file along with your source code files, we could have a better impression on how the addon works.
Thanks for your great work sofar.

**Comment by cary-rowen** [comment]
Hi,

Can NV Access take a look at this OCR project?
It has been used in some screen readers in China (including mobile screen readers), and according to my tests, it works well.

https://github.com/PaddlePaddle/PaddleOCR

**Comment by jcsteh** [comment]
I'd suggest filing a separate enhancement request issue for that.
Continuous OCR is independent of the OCR engine.

**Comment by cary-rowen** [comment]
OK, I will do it.

**Comment by Adriani90** [comment]
The new whisper.cpp from Open AI is quite successful in creating subtitles in many languages from Videos and Audios, and it might be really interesting to look in the potential for NVDA, becoming more independent from external sources such as OCR or automatic generated subtitles from Youtube etc. Here is a GUI to use the whisper.cpp in Python.
https://www.reddit.com/r/Python/comments/12kyfl4/i_made_a_simple_gui_to_use_whispercpp_in_python/

cc: @jcsteh, @leonardder

**Comment by Adriani90** [comment]
The GUI seems to work offline without any API, so I wonder if this would even work without internet connection?
